File: ui/src/lib/scopedContracts.ts
/** Scoped contracts only: transport adapters must parse untrusted JSON. */
export type Epoch = string;
export interface TargetRef {
  readonly hostId: string;
  readonly ownerId: string;
  /** Canonical decimal u64, never a JavaScript number. */
  readonly epoch: Epoch;
  readonly backendSessionId: string;
}

export type RunTarget = { readonly kind: "local" } | { readonly kind: "ssh"; readonly hostId: string };
export type ScopeCapability = "scopeControlV1" | "sshHelperV1" | "managedCodexV1" | "captureV1";
export type InventoryCompleteness = "complete" | "partial" | "unknown";
export interface InventorySnapshot<T> {
  readonly revision: number;
  readonly items: readonly T[];
  readonly completeness: InventoryCompleteness;
  readonly unavailableHosts: readonly string[];
}
export type CanonicalProvider = "codex" | "claude";
export type TransitionKind = "waiting" | "working" | "idle" | "taskComplete" | "stopped" | "removed";
export type TransitionSource =
  | { readonly kind: "provider"; readonly provider: CanonicalProvider; readonly requestId: string | null }
  | { readonly kind: "terminalDetection"; readonly detector: string }
  | { readonly kind: "lifecycle" };
export interface InventoryTransition {
  readonly revision: number;
  readonly target: TargetRef;
  readonly kind: TransitionKind;
  readonly source: TransitionSource;
}
export interface ConversationClaimKey {
  readonly hostId: string;
  readonly provider: CanonicalProvider;
  readonly conversationId: string;
}
export type ConversationOwner =
  | { readonly kind: "native"; readonly target: TargetRef }
  | { readonly kind: "managed"; readonly target: TargetRef }
  | { readonly kind: "provisional"; readonly requestId: string };
export interface ControlLease {
  readonly target: TargetRef;
  readonly deviceId: string;
  readonly leaseId: string;
  readonly expiresAtMs: number;
}
export interface MutationEnvelope<P> {
  readonly requestId: string;
  readonly target?: TargetRef;
  readonly params: P;
}
export type ScopeErrorCode = "INVALID_REQUEST" | "UNAUTHORIZED" | "FORBIDDEN" | "NOT_FOUND"
  | "TARGET_EXPIRED" | "CONTROL_CONFLICT" | "REQUEST_CONFLICT" | "PROVIDER_OWNED"
  | "UNSUPPORTED" | "TIMEOUT" | "INVENTORY_INCOMPLETE" | "PAYLOAD_TOO_LARGE" | "CAPTURE_UNSUPPORTED";
export type JsonValue = null | boolean | number | string | readonly JsonValue[] | { readonly [key: string]: JsonValue };
export interface ScopeError<D = JsonValue> {
  readonly code: ScopeErrorCode;
  readonly message: string;
  readonly retryable: boolean;
  readonly details: D;
}
export type ScopeResult<T, D = JsonValue> =
  | { readonly ok: true; readonly data: T; readonly requestId: string }
  | { readonly ok: false; readonly error: ScopeError<D>; readonly requestId: string };
export interface ScopeEvent<T> {
  readonly sequence: number;
  readonly target: TargetRef;
  readonly revision: number;
  readonly type: string;
  readonly data: T;
}
export type EventReplay<T, S> =
  | { readonly kind: "events"; readonly events: readonly ScopeEvent<T>[]; readonly afterSequence: number }
  | { readonly kind: "gap"; readonly snapshot: InventorySnapshot<S>; readonly afterSequence: number };
export const ATTACHMENT_MAX_FILE_BYTES = 10 * 1024 * 1024;
export const ATTACHMENT_MAX_FILES_PER_TURN = 4;
export const ATTACHMENT_MAX_TURN_BYTES = 20 * 1024 * 1024;
export const ATTACHMENT_UNREFERENCED_TTL_MS = 24 * 60 * 60 * 1000;
export type AttachmentMediaType = "image/png" | "image/jpeg" | "image/webp" | "text/plain" | "application/pdf";
export interface AttachmentReceipt {
  readonly hostId: string;
  readonly attachmentId: string;
  readonly sha256: string;
  readonly sizeBytes: number;
  readonly mediaType: AttachmentMediaType;
}
export interface ChatDraft {
  readonly text: string;
  readonly attachments: readonly AttachmentReceipt[];
}
export interface ConfirmedDraft {
  readonly target: TargetRef;
  readonly browserGeneration: Epoch;
  readonly attachments: readonly AttachmentReceipt[];
  readonly note: string;
}
export type DeliveryStage = "staged" | "accepted" | "providerRead";
export interface DeliveryReceipt {
  readonly requestId: string;
  readonly target: TargetRef;
  readonly stage: DeliveryStage;
}

File: src-tauri/src/scoped_contracts.rs
//! Frozen section-2 boundary DTOs. No daemon, provider, or desktop implementation.
use serde::{Deserialize, Serialize};

/// Runtime replacement epoch; the scoped wire uses canonical decimal strings.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Epoch(pub u64);

impl Serialize for Epoch {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(&self.0.to_string())
    }
}
impl<'de> Deserialize<'de> for Epoch {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let text = String::deserialize(deserializer)?;
        let value = text.parse::<u64>().map_err(serde::de::Error::custom)?;
        if value.to_string() != text {
            return Err(serde::de::Error::custom("epoch must be canonical decimal u64"));
        }
        Ok(Self(value))
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TargetRef {
    pub host_id: String,
    pub owner_id: String,
    pub epoch: Epoch,
    pub backend_session_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum RunTarget {
    #[default]
    Local,
    Ssh { #[serde(rename = "hostId")] host_id: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ScopeCapability { ScopeControlV1, SshHelperV1, ManagedCodexV1, CaptureV1 }

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum InventoryCompleteness { Complete, Partial, Unknown }

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InventorySnapshot<T> {
    pub revision: u64,
    pub items: Vec<T>,
    pub completeness: InventoryCompleteness,
    pub unavailable_hosts: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TransitionKind { Waiting, Working, Idle, TaskComplete, Stopped, Removed }

/// Waiting carries provenance; only provider-confirmed completion uses TaskComplete.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum TransitionSource {
    Provider { provider: CanonicalProvider, request_id: Option<String> },
    TerminalDetection { detector: String },
    Lifecycle,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InventoryTransition {
    pub revision: u64,
    pub target: TargetRef,
    pub kind: TransitionKind,
    pub source: TransitionSource,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CanonicalProvider { Codex, Claude }

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConversationClaimKey {
    pub host_id: String,
    pub provider: CanonicalProvider,
    pub conversation_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum ConversationOwner {
    Native { target: TargetRef },
    Managed { target: TargetRef },
    Provisional { request_id: String },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ControlLease {
    pub target: TargetRef,
    pub device_id: String,
    pub lease_id: String,
    pub expires_at_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MutationEnvelope<P> {
    pub request_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target: Option<TargetRef>,
    pub params: P,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ScopeErrorCode {
    InvalidRequest, Unauthorized, Forbidden, NotFound, TargetExpired,
    ControlConflict, RequestConflict, ProviderOwned, Unsupported, Timeout,
    InventoryIncomplete, PayloadTooLarge, CaptureUnsupported,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScopeError<D = serde_json::Value> {
    pub code: ScopeErrorCode,
    pub message: String,
    pub retryable: bool,
    pub details: D,
}

/// Literal bool marker makes mismatched success/error envelopes unrepresentable.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct WireBool<const VALUE: bool>;
impl<const VALUE: bool> Serialize for WireBool<VALUE> {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_bool(VALUE)
    }
}
impl<'de, const VALUE: bool> Deserialize<'de> for WireBool<VALUE> {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        if bool::deserialize(deserializer)? != VALUE {
            return Err(serde::de::Error::custom("incorrect result discriminant"));
        }
        Ok(Self)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged, deny_unknown_fields)]
pub enum ScopeResult<T, D = serde_json::Value> {
    Success { ok: WireBool<true>, data: T, #[serde(rename = "requestId")] request_id: String },
    Failure { ok: WireBool<false>, error: ScopeError<D>, #[serde(rename = "requestId")] request_id: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScopeEvent<T> {
    pub sequence: u64,
    pub target: TargetRef,
    pub revision: u64,
    #[serde(rename = "type")]
    pub event_type: String,
    pub data: T,
}

/// Subscription acquisition must pair this snapshot/cursor atomically.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum EventReplay<T, S> {
    Events { events: Vec<ScopeEvent<T>>, after_sequence: u64 },
    Gap { snapshot: InventorySnapshot<S>, after_sequence: u64 },
}

pub const ATTACHMENT_MAX_FILE_BYTES: u64 = 10 * 1024 * 1024;
pub const ATTACHMENT_MAX_FILES_PER_TURN: usize = 4;
pub const ATTACHMENT_MAX_TURN_BYTES: u64 = 20 * 1024 * 1024;
pub const ATTACHMENT_UNREFERENCED_TTL_MS: u64 = 24 * 60 * 60 * 1000;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AttachmentMediaType {
    #[serde(rename = "image/png")] Png,
    #[serde(rename = "image/jpeg")] Jpeg,
    #[serde(rename = "image/webp")] Webp,
    #[serde(rename = "text/plain")] Text,
    #[serde(rename = "application/pdf")] Pdf,
}

/// Host-private opaque receipt; intentionally contains no filesystem path.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AttachmentReceipt {
    pub host_id: String,
    pub attachment_id: String,
    pub sha256: String,
    pub size_bytes: u64,
    pub media_type: AttachmentMediaType,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatDraft {
    pub text: String,
    pub attachments: Vec<AttachmentReceipt>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfirmedDraft {
    pub target: TargetRef,
    pub browser_generation: Epoch,
    pub attachments: Vec<AttachmentReceipt>,
    pub note: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum DeliveryStage { Staged, Accepted, ProviderRead }

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeliveryReceipt {
    pub request_id: String,
    pub target: TargetRef,
    pub stage: DeliveryStage,
}

#[cfg(test)]
#[path = "scoped_contracts_tests.rs"]
mod tests;
